pkg/apis/ceph.rook.io/v1/placement.go

```diff
@@ -130,7 +141,27 @@ func GetArbiterPlacement(p PlacementSpec) Placement {
 	return p[KeyMonArbiter]
 }
 
+func (p Placement) MergeOsd(with Placement) Placement {
+	var ret Placement
+	if with.PodAffinity != nil {
+		ret.PodAffinity = with.PodAffinity
+	}
+	if with.PodAntiAffinity != nil {
+		ret.PodAntiAffinity = with.PodAntiAffinity
+	}
+	if with.TopologySpreadConstraints != nil {
+		ret.TopologySpreadConstraints = with.TopologySpreadConstraints
+	}
+	if with.Tolerations != nil || p.Tolerations != nil {
+		ret.Tolerations = p.mergeTolerations(with.Tolerations)
+	}
+	if with.NodeAffinity != nil {
+		ret.NodeAffinity = with.NodeAffinity
+	}
+	return ret
+}
```

Comment:
I have to make this change as non PVC case it was not working, this part is causing issue https://github.com/rook/rook/blob/856562748431935bf2b60b45685f871297269e1d/pkg/operator/ceph/cluster/osd/create.go#L293

it was showing 
```console
2021-08-20 05:58:55.316737 I | op-osd: 0 of the 5 storage nodes are valid
2021-08-20 05:58:55.322461 W | op-osd: no valid nodes available to run osds on nodes in namespace "rook-ceph"
```
when I had the following configuration on cluster.yaml i.e placement on both `placement.all` and `placement.osd`
```yaml
spec:
  placement:
    all:
      tolerations:
      - key: "key1"
        operator: "Equal"
        value: "value1"
        effect: "NoSchedule"
    osd:
      nodeAffinity:
        requiredDuringSchedulingIgnoredDuringExecution:
          nodeSelectorTerms:
          - matchExpressions:
            - key: disktype
              operator: In
              values:
              - ssd
      tolerations:
      - key: "key2"
        operator: "Equal"
        value: "value1"
        effect: "NoSchedule"```

Reply:
What was the osd spec that you got for above settings?

Reply:
What was the osd spec that you got for above settings?

Reply:
with current changes, I'm getting this [comment](https://github.com/rook/rook/pull/8566#issuecomment-902552436) below.

Reply:
which is the expected result

Reply:
I don't see any reason why `p.All().Merge(p[KeyOSD])` should not work.  It replaces the toleration added to `placement.All` with `placement.osd`.  The new method `p.All().MergeOSD(p[KeyOSD])` is just appending both the tolerations.  

Can you confirm again if tolerations applied in  `spec.placement.osd` in above example match with `taints` you have added on the nodes? 

I'm guessing you wanted to add `spec.placement.osd.tolerations[0].value` as `value2`, instead of `value1`

Reply:
> I don't see any reason why `p.All().Merge(p[KeyOSD])` should not work. It replaces the toleration added to `placement.All` with `placement.osd`. The new method `p.All().MergeOSD(p[KeyOSD])` is just appending both the tolerations.
> 
right, this replaces toleration which we don't want to append the tolerations that's why I'm using the new method `p.All().MergeOSD(p[KeyOSD])`  to append similar approach to nodeAffinity.
> Can you confirm again if tolerations applied in `spec.placement.osd` in above example match with `taints` you have added on the nodes?

> I'm guessing you wanted to add `spec.placement.osd.tolerations[0].value` as `value2`, instead of `value1`